### ISSUE TITLE
Detect unused private fields in classes

### ISSUE THREAD

**Issue by TimvdLippe** [eslint/eslint#14859]
**What rule do you want to change?**
`no-unused-vars` (arguably this could be a separate rule as well, something like `no-unused-private-fields`)
**Does this change cause the rule to produce more or fewer warnings?**
More
**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior
**Please provide some example code that this change will affect:**

```js
class MyClass {
  #unusedField = 5;
  #usedFieldOnlyInWrite = 42;
  #usedWhenReadingAndWriting = 500;
  someMethod() {
    this.#usedFieldOnlyInWrite = 42;
    this.#usedWhenReadingAndWriting += 1;
  }
  someOtherMethod() {
    return this.#usedWhenReadingAndWriting;
  }
}
```

**What does the rule currently do for this code?**
Nothing
**What will the rule do after it's changed?**
Report unused fields declared in a class. In the above example, it should report that both `#unusedField` and `#usedFieldOnlyInWrite` are unused. This should be possible, as private fields are only concerned in a particular class context, so you can collect all fields in a class, iterate through all usages and mark does that are never read as unused.
**Are you willing to submit a pull request to implement this change?**
I would be able to give it a go, but I am not sure how difficult this would be.

**Comment by nzakas** [comment]
I don't think we want to update `no-unused-vars` to cover class fields, as this is two steps beyond the intent of the rule and doesn't really apply the same logic of checking scopes (which we use to determine unused variables).

I could see the utility of a new rule, `no-unused-private-fields`, that just checks private fields for usage. Would that solve your use case?

**Comment by TimvdLippe** [comment]
Yes a new rule would be absolutely fine by me. Would you be okay if I submit a PR with that?

**Comment by nzakas** [comment]
I will champion this as a new rule. We still need to get feedback from the team so please hold off on a PR until we mark this as accepted.

**Comment by mdjermanovic** [comment]
Makes sense to add this rule. If we want to include private methods in the same rule (I don't see a reason why we shouldn't), a name like `no-unused-private-class-members` might be more appropriate.

**Comment by snitin315** [comment]
Makes sense to add this rule 👍

**Comment by nzakas** [comment]
> a name like `no-unused-private-class-members` might be more appropriate.

Makes sense to me.

**Comment by TimvdLippe** [comment]
I know this issue hasn't been marked as accepted yet. However, given the positive response thus far and me having a Friday to work on open source contributions, I figured I can implement the rule relatively straightforwardly (#14895). If this rule ends up not being accepted, I am planning to copy it into our DevTools rules anyways. So the work won't be thrown away for me personally.

P.S. What are the odds that this issue has number 14859 and the PR I made number 14895. Made me chuckle 😄

**Comment by nzakas** [comment]
I'm reasonably certain this will be accepted, it just takes time to get in front of more of the team.

**Comment by TimvdLippe** [comment]
I included this rule in DevTools today and it found a whole bunch of unused fields :tada: https://crrev.com/c/3310610 I didn't encounter any further issues nor false positives, so I think we are good here :smile:
